config/credential_handler.rb

```diff
@@ -32,15 +32,25 @@ def self.call(request)
   # This requires the use client_credentials grant flow with JWT assertions.
   # This method handles authorizing the client by decoding the JWT assertion for the supported client credentials flow. 
   def self.handle_client_credentials_flow(request)
-    # Extract the shared client ID.
-    client_id = request.parameters[:client_id]
-    raise_missing_param_error('client_id') unless client_id.present?
-
     # Extract signed JWT assertion from client_assertion param.
     # This is what is signed with the client secret and will be used to authorize this request.
     signed_token = request.parameters[:client_assertion]
     raise_missing_param_error('client_assertion') unless signed_token.present?
 
+    # Extract shared client ID.
+    if request.parameters[:client_id].present?
+      # Check to see if client_id included in params (optional)
+      client_id = request.parameters[:client_id]
+    else
+      # If not, decode the signed token first without validation to obtain the client ID.
+      decoded_assertion = JWT.decode(signed_token, nil, false)
+      # Grab the JWT payload from the decode output array
+      payload = decoded_assertion.length > 0 ? decoded_assertion[0] : nil
+      # Both the iss and sub fields should contain the client ID.
+      client_id = payload["iss"] ? payload["iss"] : payload["sub"]
```

Comment:
if payload is `nil` then `payload["iss"]` will barf.

Reply:
Good catch, fixed.